src/fromager/vendor_rust.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/python3
+"""Vendor Rust crates into an sdist
+"""
+import os
+import pathlib
+import subprocess
+import tarfile
+import tempfile
+import typing
+
+from packaging.version import Version
+import toml
+
+# minimum supported Rust version
+# RHEL 9.4 has rustc 1.75
+RHEL9_MSRV = Version("1.75")
+
+VENDOR_DIR = "vendor"
+
+# Windows, macOS, and iOS archives and libs
+VENDOR_FILTER_PATTERNS = ("**/windows*/**/*.a", "**/windows*/**/*.lib")
+
+CARGO_CONFIG = {
+    "source": {
+        "crates-io": {"replace-with": "vendored-sources"},
+        "vendored-sources": {"directory": VENDOR_DIR},
+    },
+}
+
+
+class NoCargoToml(Exception):
+    """Source dist has no Cargo.toml files"""
+
+
+def _tar_reset(tarinfo: tarfile.TarInfo) -> tarfile.TarInfo:
+    """Reset user, group, mtime, and mode to create reproducible tar"""
+    tarinfo.uid = 0
+    tarinfo.gid = 0
+    tarinfo.uname = "root"
+    tarinfo.gname = "root"
+    tarinfo.mtime = 0
+    if tarinfo.type == tarfile.DIRTYPE:
+        tarinfo.mode = 0o755
+    else:
+        tarinfo.mode = 0o644
+    if tarinfo.pax_headers:
+        raise ValueError(tarinfo.name, tarinfo.pax_headers)
+    return tarinfo
+
+
+def _tar_reproducible(tar: tarfile.TarFile, basedir: pathlib.Path) -> None:
+    """Create reproducible tar file"""
+    content = []
+    for root, dirs, files in basedir.walk():
+        for directory in dirs:
+            content.append(root / directory)
+        for filename in files:
+            content.append(root / filename)
+
+    content.sort()
+
+    for fn in content:
+        tar.add(
+            fn,
+            filter=_tar_reset,
+            recursive=False,
+            arcname=fn.relative_to(basedir),
+        )
+
+
+def _cargo_vendor(
+    manifests: typing.List[pathlib.Path],
+    project_dir: pathlib.Path,
+) -> None:
+    """Run cargo vendor"""
+    args = ["cargo", "vendor", f"--manifest-path={manifests[0]}"]
+    for manifest in manifests[1:]:
+        args.append(f"--sync={manifest}")
+    args.append(project_dir / VENDOR_DIR)
+    subprocess.check_call(args, stdout=subprocess.DEVNULL, env={})
+
+    # Remove large, unused archive and lib files. They are only
+    # used on Windows, macOS, and iOS. This makes the vendor bundle up to
+    # 60% smaller.
+    for pattern in VENDOR_FILTER_PATTERNS:
+        for filename in project_dir.glob(pattern):
+            filename.unlink()
+            filename.touch()  # create empty file
+
+
+def _fixup_cargo_manifest(
+    fileame: pathlib.Path,
+    *,
+    msrv: typing.Optional[Version] = None,
+):
+    """Modify Cargo.toml
+
+    pydantic-core has an unnecessary high MSRV.
+    """
+    with open(fileame, "r", encoding="utf-8") as f:
+        cfg = toml.load(f)
+
+    modified = False
```

Comment:
I'm trying to keep project-specific knowledge out of the core bits, and provide an override hook or other mechanism for a build system to inject that sort of knowledge based on the packages it wants to build. In this case, for setting the rust level, I could see us setting a global config in the settings file, or I could see a patch being applied to the package.